### PR TITLE
add `PianoContext`

### PR DESCRIPTION
**Pull request by thisjiang**
### PR types

New features
### PR changes

Others
### Describe

> **WARNING:** To compile and run this PR, you need merge the previous PR: https://github.com/wzzju/Paddle/pull/19

This PR used to define the structure of `PianoScope` and `PianoContext`.

# `PianoScope`
Like `Scope` class in Paddle, the main effect  of `PianoScope` class is map the `operand` name to its `operand` value. The same as `Scope`, `PianoScope` has ancestor and child scope, we can invoke `NewScope` function to create new child scope, the child scope can access the `operand` in its  ancestor scope, but not vice versa.

# `PianoContext`
Generally, The `PianoContext` is passing to the `PianoOpKernel::Compile` funtion, but unlike `ExecutionContext` class in Paddle, the structure of `PianoContext` is more simple. It's constructed function accept three parameter: `OpDesc`, `PianoScope` and `NoteBuilder`.
```
PianoContext(const framework::OpDesc* op_desc, PianoScope* scope, NoteBuilder* builder)
``` 
The `OpDesc` is used to obtain the op information like input, output, and paddle attribute. Besides, we also need the piano `OpRegistration` information, so **the op need registered though `REGISTER_PIANO_OP` before!**

The `PianoScope` store the map from name to `Operand`.

## Descriptionn
In `PianoOpKernel::Compile` funtion, developer deal with `Operand` rather than `Variable`, the developer obtain the input `Operand` by `Input` function and should create output `Operand` which is generated by some note IR operation.
```
Operand* Input(const std::string& name) const;  // return input Operand
void SetOutput(const std::string& name, Operand* op) const;  // set output Operand 
```
**Note that the input and output name must need be founded in `OpDesc`'s input and output.**

As mentioned above, except the attributes in `OpDesc`, the `PianoContext` also has its Piano attributes, and the priority of piano attribute is higher than origin's, so the `GetAttr` funtion will return the Piano attribute value if the attribute name is same.

# Single Test
image

### CODE REVIEW

File: paddle/fluid/compiler/paddle2piano/piano_context.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+namespace paddle {
+namespace piano {
+
+// The PianoContext used to transmit the higher level information to the
+// lower level of piano note IR.
+// "OpDesc" is the operator information.
+// "PianoScope" is an association of a name to operand.
+// "builder" is the operand's NoteBuilder.
+class PianoContext {
```
Comment (CtfGo) [inline paddle/fluid/compiler/paddle2piano/piano_context.h]: 建议类名为PianoOpKernelContext或PianoOpContext。PianoContext名称限定范围太大，可以用在整个Piano编译执行阶段的上下文定义

File: paddle/fluid/compiler/paddle2piano/piano_context.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ public:
+  PianoContext(const framework::OpDesc* op_desc, PianoScope* scope,
+               NoteBuilder* builder)
+      : op_(op_desc),
+        scope_(scope),
+        builder_(builder),
+        op_reg_(PianoOpRegistry::Instance().ops_.at(op_desc->Type()).get()) {}
+
+  const std::string Type() const { return op_->Type(); }
```
Comment (CtfGo) [inline paddle/fluid/compiler/paddle2piano/piano_context.h]: 加引用

File: paddle/fluid/compiler/paddle2piano/piano_context.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  Operand* Input(const std::string& name) const;
+
+  // Map the outputs's operand into scope, the operand is created by
+  // NoteBuilder, and be careful the output name must existed in op's
+  // outputs.
+  void SetOutput(const std::string& name, Operand* op) const;
+
+  // the piano op attributes
+  bool HasAllowBackendList() const { return op_reg_->has_allow_backend_list; }
```
Comment (CtfGo) [inline paddle/fluid/compiler/paddle2piano/piano_context.h]: 这个属性kernel里应该不会用到？目前想到是应用在编译开始前判断op是否支持piano（graph pass搜索替换时）

File: paddle/fluid/compiler/paddle2piano/piano_context.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  std::vector<std::string> AllowBackendList() const {
+    return HasAllowBackendList() ? op_reg_->allow_backend_list
+                                 : PianoOpRegistry::AllBackendNames();
+  }
+
+  std::vector<Layout> Layouts() const { return op_reg_->supported_layouts; }
```
Comment (CtfGo) [inline paddle/fluid/compiler/paddle2piano/piano_context.h]: 这里的layout是框架的吗？piano里也有个paddle::piano::Layout

File: paddle/fluid/compiler/paddle2piano/piano_context_test.cc [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  }
+};
+
+// register piano op kernel with limit allow backend list
+class TestPianoOpMaker : public PianoOpMaker {
+ public:
+  void Make() override {
+    SetAllowBackendList({"PIANO_JIT_TEST_TYPE"});
+    SetDataTypes(TestDatatypes());
+    SetLayouts({Layout({3, 1, 0, 2})});
```
Comment (CtfGo) [inline paddle/fluid/compiler/paddle2piano/piano_context_test.cc]: PianoOpRegistry里包含这个layout信息是处于什么考虑呢

File: paddle/fluid/compiler/paddle2piano/piano_scope.h [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    }
+    return ret;
+  }
+
+ private:
+  explicit PianoScope(const PianoScope* parent) : parent_(parent) {}
+
+  DISABLE_COPY_AND_ASSIGN(PianoScope);
+
+  std::unordered_map<std::string, Operand*> operands_;
```
Comment (CtfGo) [inline paddle/fluid/compiler/paddle2piano/piano_scope.h]: 这里不能只存指针，元算子返回的Operand是个栈数据，可以直接拷贝存储，Opearnd里只有int和builder指针两个元素

File: paddle/fluid/compiler/paddle2piano/piano_context.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+  const std::string Type() const { return op_->Type(); }
+
+  NoteBuilder* Builder() const { return builder_; }
+
+  PianoScope* Scope() const { return scope_; }
+
+  const framework::OpDesc* Op() const { return op_; }
+
+  std::vector<std::string> InputNames() const { return op_->InputNames(); }
```
Comment (CtfGo) [inline paddle/fluid/compiler/paddle2piano/piano_context.h]: Scope、Op、InputNames、OutputNames接口感觉kernel里暂时没有看到需要用，是不是可以先不加，后面有需要再补，尽量保持初始接口精简一些。

Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_context.h]: Done

File: paddle/fluid/compiler/paddle2piano/piano_scope.h (hunk 6 above)
Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_scope.h]: Done

File: paddle/fluid/compiler/paddle2piano/piano_context_test.cc (hunk 5 above)
Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_context_test.cc]: 应该是框架原有的`framework::DataLayout`，我看原来Paddle OpKernel注册是是有这个信息的，Piano就也加上了

File: paddle/fluid/compiler/paddle2piano/piano_context.h (hunk 4 above)
Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_context.h]: 对的，是框架原有的`framework::DataLayout`，我改下

File: paddle/fluid/compiler/paddle2piano/piano_context.h (hunk 3 above)
Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_context.h]: 行，那我删了

File: paddle/fluid/compiler/paddle2piano/piano_context.h (hunk 2 above)
Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_context.h]: Done

File: paddle/fluid/compiler/paddle2piano/piano_context.h (hunk 1 above)
Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_context.h]: Done

File: paddle/fluid/compiler/paddle2piano/CMakeLists.txt [hunk 8]
```diff
@@ -1,2 +1,8 @@
 cc_library(piano_compile_pass SRCS piano_compile_pass.cc DEPS pass subgraph_detector)
 cc_test(piano_compile_pass_test SRCS piano_compile_pass_tester.cc DEPS piano_compile_pass)
+
+cc_library(piano_op_registry SRCS piano_op_registry.cc DEPS framework_proto note_proto piano_data_description)
+cc_test(piano_op_registry_test SRCS piano_op_registry_test.cc DEPS piano_op_registry operator op_registry)
+
+cc_library(piano_context SRCS piano_context.cc DEPS piano_op_registry proto_desc note_builder)
```
Comment (CtfGo) [inline paddle/fluid/compiler/paddle2piano/CMakeLists.txt]: 文件名相应改一下？

File: paddle/fluid/compiler/paddle2piano/piano_op_kernel_context.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  framework::Attribute GetAttr(const std::string& name) const;
+
+  template <typename T>
+  T GetAttr(const std::string& name) const {
+    return BOOST_GET_CONST(T, GetAttr(name));
+  }
+
+ private:
+  const framework::OpDesc* op_;
+  mutable PianoScope* scope_;
```
Comment (levi131) [inline paddle/fluid/compiler/paddle2piano/piano_op_kernel_context.h:74]: 这里为什么使用 mutable呀

File: paddle/fluid/compiler/paddle2piano/piano_op_kernel_context.h [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  // cannot returning reference to temporary
+  const std::string Type() const { return op_->Type(); }
+
+  NoteBuilder* Builder() const { return builder_; }
+
+  bool HasInput(const std::string& name) const {
+    return op_->Inputs().find(name) != op_->Inputs().end();
+  }
+
+  Operand Input(const std::string& name) const;
```
Comment (levi131) [inline paddle/fluid/compiler/paddle2piano/piano_op_kernel_context.h]: GetInput() 和 SetOutput()是不是更加搭配一点

File: paddle/fluid/compiler/paddle2piano/piano_op_kernel_context.h (hunk 9 above)
Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_op_kernel_context.h:74]: 因为考虑到传入到`PianoOpKernel`的参数是`const PianoOpKernelContext&`，如果不设置为mutable的话那我们就不能在`SetOutput`中修改`scope_`的值了。

File: paddle/fluid/compiler/paddle2piano/piano_op_kernel_context.h (hunk 10 above)
Comment (thisjiang) [reply paddle/fluid/compiler/paddle2piano/piano_op_kernel_context.h]: 好主意，我改下

**Review by levi131 — APPROVED** [review]
LGTM

**Review by CtfGo — APPROVED** [review]
LGTM

**Pull request merged**